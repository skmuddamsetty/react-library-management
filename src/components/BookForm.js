import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class BookForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.book ? props.book.title : '',
      description: props.book ? props.book.description : '',
      price: props.book ? (props.book.price / 100).toString() : '',
      createdAt: props.book ? moment(props.book.publishedAt) : moment(),
      calendarFocused: false,
      publisher: props.book ? props.book.publisher : '',
      error: '',
    };
  }

  titleChangeHandler = (e) => {
    const title = e.target.value;
    this.setState(() => ({
      title,
    }));
  };
  descriptionChangeHandler = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description,
    }));
  };
  priceChangeHandler = (e) => {
    const price = e.target.value;
    if (!price || price.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        price,
      }));
    }
  };
  dateChangeHandler = (createdAt) => {
    if (createdAt) {
      this.setState(() => {
        createdAt;
      });
    }
  };
  dateFocusChangeHandler = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused,
    }));
  };
  publisherChangeHandler = (e) => {
    const value = e.target.value;
    this.setState(() => ({ publisher: value }));
  };
  submitHandler = (e) => {
    e.preventDefault();
    if (!this.state.title || !this.state.price) {
      this.setState(() => ({
        error: 'Please provide title and price for the book!',
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        title: this.state.title,
        description: this.state.description,
        price: parseFloat(this.state.price, 10) * 100, // converting to cents by multiplying with 100
        publishedAt: moment(this.state.createdAt).valueOf(), // since createdAt is a moment object converting it to timestamp using valueOf
      });
    }
  };
  render() {
    return (
      <form onSubmit={this.submitHandler} className='form'>
        {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input
          className='text-input'
          type='text'
          placeholder='Title'
          autoFocus
          value={this.state.title}
          onChange={this.titleChangeHandler}
        />
        <input
          className='text-input'
          type='text'
          placeholder='Publisher'
          autoFocus
          value={this.state.publisher}
          onChange={this.publisherChangeHandler}
        />
        <input
          className='text-input'
          type='text'
          placeholder='Amount'
          value={this.state.price}
          onChange={this.priceChangeHandler}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.dateChangeHandler}
          onFocusChange={this.dateFocusChangeHandler}
          focused={this.state.calendarFocused}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className='text-area'
          placeholder='Add a description for your book'
          value={this.state.description}
          onChange={this.descriptionChangeHandler}
        ></textarea>
        <div>
          <button className='button'>Save Book</button>
        </div>
      </form>
    );
  }
}

export default BookForm;
