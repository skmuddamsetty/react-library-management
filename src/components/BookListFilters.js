import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByPrice } from '../actions/filters';
const BookListFilters = (props) => (
  <div>
    <input
      type='text'
      value={props.filters.text}
      onChange={(e) => {
        props.dispatch(setTextFilter(e.target.value));
      }}
    />
    <select
      onChange={(e) => {
        if (e.target.value === 'date') {
          props.dispatch(sortByDate());
        } else if (e.target.value === 'price') {
          props.dispatch(sortByPrice());
        }
      }}
      value={props.filters.sortBy}
    >
      <option value='price'>Price</option>
      <option value='date'>Published Date</option>
    </select>
  </div>
);

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(BookListFilters);
