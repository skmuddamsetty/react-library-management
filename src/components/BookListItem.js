import React from 'react';
import { Link } from 'react-router-dom';

const BookListItem = ({
  title,
  description,
  price,
  publishedAt,
  id,
  dispatch,
  history,
}) => (
  <div>
    <Link to={`/edit-book/${id}`}>
      <h3>
        {title}
        <span> - {description}</span>
      </h3>
    </Link>
    <p>{price}</p>
    <p>{publishedAt}</p>
  </div>
);

export default BookListItem;
