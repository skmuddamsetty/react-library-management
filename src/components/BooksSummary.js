import React from 'react';
import { connect } from 'react-redux';
import getVisibleBooks from '../selectors/books';
import selectBooksTotal from '../selectors/books-total';
import numeral from 'numeral';

export const BooksSummary = ({ booksCount, booksTotal }) => {
  const bookWord = booksCount === 1 ? 'book' : 'books';
  const formattedBooksTotal = numeral(booksTotal / 100).format('$0,0.00');
  return (
    <h2>
      Viewing {booksCount} {bookWord} totalling {formattedBooksTotal}
    </h2>
  );
};

const mapStateToProps = (state) => {
  const visibleBooks = getVisibleBooks(state.books, state.filters);
  return {
    booksCount: visibleBooks.length,
    booksTotal: selectBooksTotal(visibleBooks),
  };
};

export default connect(mapStateToProps)(BooksSummary);
