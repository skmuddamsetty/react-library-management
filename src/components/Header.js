import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = (props) => (
  <header className='header'>
    <div className='content-container'>
      <div className='header__content'>
        <Link to='/dashboard' className='header__title'>
          <h1>Library Management</h1>
        </Link>
        <button className='button button--link' onClick={props.startLogout}>
          Logout
        </button>
      </div>
    </div>
  </header>
);

const mapPropsToDispatch = (dispatch) => {
  return {
    startLogout: () => dispatch(startLogout()),
  };
};
export default connect(undefined, mapPropsToDispatch)(Header);
