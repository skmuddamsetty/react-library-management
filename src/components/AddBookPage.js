import React from 'react';
import BookForm from './BookForm';
import { connect } from 'react-redux';
import { startAddBook } from '../actions/books';

export class AddBookPage extends React.Component {
  onSubmit = (book) => {
    // we should not use the below line as this makes testing this component harder
    // props.dispatch(addBook(book));
    // we are calling the addBook property which has been obtained from mapDispatchToProps
    this.props.startAddBook(book);
    // navigating the user to the dashboard page
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Add Book</h1>
          </div>
        </div>
        <div className='content-container'>
          <BookForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}
// const AddBookPage = (props) => (
//   <div>
//     <h1>Add Book Page</h1>
//     <BookForm
//       onSubmit={(book) => {
//         // we should not use the below line as this makes testing this component harder
//         // props.dispatch(addBook(book));
//         // we are calling the onSubmit property which has been obtained from mapDispatchToProps
//         props.onSubmit(book);
//         // navigating the user to the dashboard page
//         props.history.push('/');
//       }}
//     />
//   </div>
// );

// const mapDispatchToProps = (dispatch) => {
//   return {
//     onSubmit: (book) => dispatch(addBook(book)),
//   };
// };

// syntax for implicitly returning the object. this is same as the above commented code but uses a short hand syntax
const mapDispatchToProps = (dispatch) => ({
  startAddBook: (book) => dispatch(startAddBook(book)),
});
export default connect(undefined, mapDispatchToProps)(AddBookPage);
