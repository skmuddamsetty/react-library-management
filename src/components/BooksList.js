import React from 'react';
import { connect } from 'react-redux';
import BookListItem from './BookListItem';
import getVisibleBooks from '../selectors/books';

export const BooksList = (props) => (
  <div>
    {props.books.length === 0 ? (
      <p>No Books</p>
    ) : (
      props.books.map((book) => {
        return <BookListItem key={book.id} {...book} />;
      })
    )}
  </div>
);

// const ConnectedBooksList = connect((state) => {
//   return {
//     books: state.books,
//   };
// })(BooksList);

const mapStateToProps = (state) => {
  return {
    books: getVisibleBooks(state.books, state.filters),
  };
};

export default connect(mapStateToProps)(BooksList);
