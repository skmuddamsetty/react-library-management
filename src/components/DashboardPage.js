import React from 'react';
import BooksList from './BooksList';
import BookListFilters from './BookListFilters';
const LibraryDashboardPage = () => (
  <div>
    <BookListFilters />
    <BooksList />
  </div>
);
export default LibraryDashboardPage;
