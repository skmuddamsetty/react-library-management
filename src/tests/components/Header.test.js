// import React from 'react';
// import ReactShallowRenderer from 'react-test-renderer/shallow';
// import Header from '../../components/Header';

// test('should render Header correctly', () => {
//   const renderer = new ReactShallowRenderer();
//   renderer.render(<Header />);
//   expect(renderer.getRenderOutput()).toMatchSnapshot();
// });
import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

test('should render Header correctly', () => {
  const wrapper = shallow(<Header startLogout={() => {}} />);
  expect(wrapper.find('h1').length).toBe(1);
  expect(wrapper.find('h1').text()).toBe('Library Management');
});

test('should render Header correctly based off of snapshot', () => {
  const wrapper = shallow(<Header startLogout={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('should logout on button click', () => {
  const startLogout = jest.fn();
  const wrapper = shallow(<Header startLogout={startLogout} />);
  wrapper.find('button').simulate('click');
  expect(startLogout).toHaveBeenCalled();
});
