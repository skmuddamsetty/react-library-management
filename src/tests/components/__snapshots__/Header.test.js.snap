// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly based off of snapshot 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        to="/"
      >
        <h1>
          Library Management
        </h1>
      </Link>
      <button
        className="button button--link"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
